test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -993,5 +993,67 @@ def test_hsdp_broadcast_across_replicas(self):
         model(inp).sum().backward()
 
 
+class TestFullyShardShardPlacementFn(FSDPTestMultiThread):
+    @property
+    def world_size(self) -> int:
+        return 8
+
+    def _init_models(self):
+        torch.manual_seed(42)
+        model_args = ModelArgs(n_layers=3, dropout_p=0.0)
+        model = Transformer(model_args)
+        for param in model.parameters():
+            dist.broadcast(param.detach(), src=0)
+        ref_model = copy.deepcopy(model)
+        return model, ref_model
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_init_1d_transformer_flat_shard(self):
+        model, ref_model = self._init_models()
+
+        def shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
+            return _FlatShard(0)
+
+        for layer in model.layers:
+            fully_shard(layer, shard_placement_fn=shard_placement_fn)
+        fully_shard(model, shard_placement_fn=shard_placement_fn)
+
+        for param, ref_param in zip(model.parameters(), ref_model.parameters()):
+            full_param = param.full_tensor()
+            # NOTE: We must view back from 1D to ND.
+            full_param = full_param.view(param.shape)
```

Comment:
for understanding purpose, DTensor.shape is original shape, but DTensor.full_tensor() is flattened shape?

Reply:
yep unfortunately it is like that right now
because `_FlatShard` is inheriting from `Shard`, the implementation to convert to replicated is the same, which will gather along the shard dim (which is 0), so it will still be 1D after the all-gather

I am not sure what the right thing to do here is exactly

Reply:
I think that the `DTensor.shape` should be the original unflattened shape since we need to hold that info somewhere. If we assume that as axiomatic, then we need to implement the ops for redistribution to be correct for `_FlatShard`, or we need to raise errors that certain redistribution is not supported.

We can fill out the `_FlatShard` redistribution logic over time. I can spend some time on it too later, but we should prioritize understanding which direction we want to go.

Reply:
Shall we do the reshape here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_redistribute.py#L156

Reply:
thanks for explaining. totally understand this is for directional exploration